src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -0,0 +1,481 @@
+		}
+
+		@Override
+		public void run() {
+			while (isRunning()) {
+				Set<Partition> partitionsWithRemainingData;
+				boolean hasErrors;
+				do {
+					MutableCollection<Partition> fetchPartitions = partitionsByBrokerMap.get(brokerAddress);
+					partitionsWithRemainingData = new HashSet<Partition>();
+					hasErrors = false;
+					try {
+						MutableCollection<FetchRequest> fetchRequests = fetchPartitions.collect(new PartitionToFetchRequestFunction());
+						Result<KafkaMessageBatch> result = kafkaTemplate.receive(fetchRequests);
+						// process successful messages first
+						Iterable<KafkaMessageBatch> batches = result.getResults().values();
+						for (KafkaMessageBatch batch : batches) {
+							if (!batch.getMessages().isEmpty()) {
+								long highestFetchedOffset = 0;
+								for (KafkaMessage kafkaMessage : batch.getMessages()) {
+									// fetch operations may return entire blocks of compressed messages, which may have lower offsets than the ones requested
+									// thus a batch may contain messages that have been processed already
+									if (kafkaMessage.getMetadata().getOffset() >= fetchOffsets.get(batch.getPartition())) {
+										messageDispatcher.dispatch(kafkaMessage);
+									}
+									highestFetchedOffset = Math.max(highestFetchedOffset, kafkaMessage.getMetadata().getNextOffset());
+								}
+								fetchOffsets.replace(batch.getPartition(), highestFetchedOffset);
+								// if there are still messages on server, we can go on and retrieve more
+								if (highestFetchedOffset < batch.getHighWatermark()) {
+									partitionsWithRemainingData.add(batch.getPartition());
+								}
+							}
+						}
+						// handle errors
+						if (result.getErrors().size() > 0) {
+							hasErrors = true;
+
+							// find partitions with leader errors and
+							PartitionIterable<Map.Entry<Partition, Short>> partitionByLeaderErrors =
+									partition(result.getErrors().entrySet(), new IsLeaderErrorPredicate());
+							RichIterable<Partition> partitionsWithLeaderErrors =
+									partitionByLeaderErrors.getSelected().collect(keyFunction);
+							resetLeaders(partitionsWithLeaderErrors);
+
+							PartitionIterable<Map.Entry<Partition, Short>> partitionsWithOffsetsOutOfRange
+									= partitionByLeaderErrors.getRejected().partition(new IsOffsetOutOfRangePredicate());
+							resetOffsets(partitionsWithOffsetsOutOfRange.getSelected().collect(keyFunction).toSet());
+							// it's not a leader issue
+							stopFetchingFromPartitions(partitionsWithOffsetsOutOfRange.getRejected().collect(keyFunction));
+						}
+
+					}
+					catch (ConsumerException e) {
+						// this is a broker error, and we cannot recover from it. Reset leaders and stop fetching data from this
+						// broker altogether
+						log.error(e);
+						resetLeaders(fetchPartitions.toImmutable());
+						return;
+					}
+				} while (!hasErrors && !partitionsWithRemainingData.isEmpty());
+				try {
+					Thread.sleep(interval);
```

Comment:
You're right I don't think we do, either. My original intent was to prevent entering a busy wait while not having any incoming data, but, obviously, the long polling on fetch would take care of that.